apps/src/pythonlab/PythonlabView.tsx

```diff
@@ -169,13 +173,41 @@ const PythonlabView: React.FunctionComponent<
   // Set view code to false if level is switched for any levels in widget view.
   useLifecycleNotifier(LifecycleEvent.LevelLoadStarted, () => {
     dispatch(setWidgetViewShowCode(false));
+    setAiTutor2HintQuestion(undefined);
   });
 
+  // Given a question for the AITutor2, return the full question to ask, which means
+  // appending all the relevant context.
+  const getAiTutor2FullQuestionFromQuestion = (
+    question: string,
+    questionType: 'hint' | 'user'
+  ) => {
+    const fullQuestion = [
+      question,
+      'Here is my code:',
+      source.openFiles
+        ?.map(openFile => source.files[openFile].contents)
+        .join('\n'),
+      'Here is the validation code:',
```

Comment:
is it a problem if the user hasn't run validation and/or there's no validation on the level?

Reply:
Good catch.  The code at least now only adds context about validation code and validation results if it has them.